### PR TITLE
PWGHF: Adding flag selection in the hadron tasks

### PR DESCRIPTION
**Pull request by DelloStritto**

### CODE REVIEW

**Review by vkucera — CHANGES_REQUESTED** [review]
If this works, it works by sheer luck.
There is currently no correspondence between the `hfflag()` values set in the index skimming and the `DecayType` values used in MC matching.
@nzardosh I suggest to use the `DecayType` values for the skimming bits as well.
Same applies to https://github.com/AliceO2Group/AliceO2/pull/5204.

**Comment by nzardosh** [comment]
> If this works, it works by sheer luck.
> There is currently no correspondence between the `hfflag()` values set in the index skimming and the `DecayType` values used in MC matching.
> @nzardosh I suggest to use the `DecayType` values for the skimming bits as well.
> Same applies to #5204.

Hi @vkucera , yes I agree. This was the idea behind Luigi's change too.

File: Analysis/Tasks/PWGHF/HFD0CandidateSelector.cxx [hunk 1]
```diff
@@ -324,11 +324,17 @@ struct HFD0CandidateSelector {
 
     for (auto& hfCandProng2 : hfCandProng2s) { //looping over 2 prong candidates
 
+      statusD0 = 0;
+      statusD0bar = 0;
+
+      if (!(hfCandProng2.hfflag() & 1 << D0ToPiK)) {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/HFD0CandidateSelector.cxx:330]: What is the operation priority here?
Is this equivalent to `if (!(hfCandProng2.hfflag() & (1 << D0ToPiK)))` or to `if (!((hfCandProng2.hfflag() & 1) << D0ToPiK))`?

File: Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h [hunk 2]
```diff
@@ -155,7 +155,8 @@ DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, int8_t); // reconstruction le
 DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t); // generator level
 
 // mapping of decay types
-enum DecayType { D0ToPiK = 1 };
+enum DecayType { D0ToPiK = 0,
+                 JPsiToEE };
```
Comment (vkucera) [inline Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h]: -> `JpsiToEE`

File: Analysis/Tasks/PWGHF/HFD0CandidateSelector.cxx (hunk 1 above)
Comment (DelloStritto) [reply Analysis/Tasks/PWGHF/HFD0CandidateSelector.cxx:330]: it's the first one. I left shift 1 by my nbit (in this case D0ToPiK)  first.

It is equivalent to what we have now but using the enumerator for the bit.
https://github.com/AliceO2Group/AliceO2/blob/dev/Analysis/Tasks/PWGHF/HFLcCandidateSelector.cxx#L292

Comment (vkucera) [reply Analysis/Tasks/PWGHF/HFD0CandidateSelector.cxx:330]: So it looks like the shift `<<` has precedence over the AND `&`.
https://en.cppreference.com/w/cpp/language/operator_precedence
So it should be fine. :-)

**Comment by vkucera** [comment]
I suppose the same bit check should be added in the rec. part of the MC tasks as well, right?

**Comment by DelloStritto** [comment]
> I suppose the same bit check should be added in the rec. part of the MC tasks as well, right?

I think it is not strictly needed because we check the bit here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Analysis/Tasks/PWGHF/taskLc.cxx#L127

**Comment by vkucera** [comment]
> > I suppose the same bit check should be added in the rec. part of the MC tasks as well, right?
> 
> I think it is not strictly needed because we check the bit here:
> https://github.com/AliceO2Group/AliceO2/blob/dev/Analysis/Tasks/PWGHF/taskLc.cxx#L127
> 
> But maybe we can add it so we do not loop over all the candidates.

I think it is needed if we want to fill the MC background histograms with the same candidates that populate the combinatorial background in the non-MC task. Isn't it?

**Comment by DelloStritto** [comment]
> > > I suppose the same bit check should be added in the rec. part of the MC tasks as well, right?
> > 
> > 
> > I think it is not strictly needed because we check the bit here:
> > https://github.com/AliceO2Group/AliceO2/blob/dev/Analysis/Tasks/PWGHF/taskLc.cxx#L127
> > But maybe we can add it so we do not loop over all the candidates.
> 
> I think it is needed if we want to fill the MC background histograms with the same candidates that populate the combinatorial background in the non-MC task. Isn't it?

I'm not sure about this. In this way we would not compute the loss in efficiency due to the skimming cuts and I'm not sure this is what we want.

I missed that you specified "rec part of the MC" in your comment. I think that you are right

**Comment by iarsene** [comment]
Hi, is this PR ready now ?

**Comment by ginnocen** [comment]
@iarsene, please wait. We have to make a few extra tests.

**Comment by iarsene** [comment]
Maybe switch to draft so we avoid mistakenly merging these PRs

**Review by ginnocen — APPROVED** [review]

**Review by iarsene — APPROVED** [review]

**Pull request merged**